NuGetUpload/Services/PackageUploadService.cs

```diff
@@ -83,94 +82,116 @@ private async IAsyncEnumerable<string> ExtractZip(string zipFilePath, string bas
+
+            var filePaths = new List<string>();
+            foreach (var file in files)
             {
-                var uploadMsg = $"Uploading {browserFile.Name}";
-                yield return Step(uploadMsg, 0);
-                var filePath = Path.Combine(randomFolder.Path, browserFile.Name);
-                var totalSize = (double)browserFile.Size;
-                await foreach (var read in Upload(browserFile, filePath))
-                    yield return Step(uploadMsg, read / totalSize * stepSize);
-
-                if (zipContentTypes.Contains(browserFile.ContentType))
+                var uploadMessage = $"Uploading {file.Name}";
+                Step(uploadMessage, 0);
+
+                var filePath = Path.Combine(temporaryFolder.Path, file.Name);
+                var totalSize = (double)file.Size;
+                await using var stream = file.OpenReadStream(MaxFileSize);
+                await foreach (var read in SaveToFile(stream, filePath))
+                    Step(uploadMessage, read / totalSize * stepSize);
+
+                if (_zipContentTypes.Contains(file.ContentType))
                 {
-                    yield return Step($"Extracting ZIP {browserFile.Name}", 0);
-                    await foreach (var s in ExtractZip(filePath, randomFolder.Path, uploads))
+                    Step($"Extracting ZIP {file.Name}", 0);
+                    await foreach (var s in ExtractZip(filePath, temporaryFolder.Path, filePaths))
                         Step(s, 0);
                 }
                 else
                 {
-                    uploads.Add(filePath);
+                    filePaths.Add(filePath);
                 }
             }
 
-            var allowedAssemblies =
-                new HashSet<string>(info.AllowedAssemblies, StringComparer.InvariantCultureIgnoreCase);
-            foreach (var filePath in uploads)
+            var allowedAssemblies = info.AllowedAssemblies == null ? null : new HashSet<string>(info.AllowedAssemblies, StringComparer.InvariantCultureIgnoreCase);
+
+            foreach (var filePath in filePaths)
             {
                 var fileName = Path.GetFileName(filePath);
-                yield return Step($"Stripping and publicising {fileName}");
                 try
                 {
-                    AssemblyStripper.StripAssembly(filePath, allowedAssemblies);
+                    using var module = ModuleDefMD.Load(await File.ReadAllBytesAsync(filePath));
+                    if (allowedAssemblies != null && !allowedAssemblies.Remove(module.Assembly.Name))
+                        throw new ArgumentException($"Assembly {module.Assembly.Name} is not in allowed list");
+
+                    if (!info.SkipStripping)
+                    {
+                        Step($"Stripping and publicising {fileName}");
+                        AssemblyStripper.StripAssembly(module);
+                        module.Write(filePath);
+                    }
```

Comment:
Any reason for moving assembly reading out of `AssemblyStripper`? I'd prefer keeping all dnlib things in a separate utility class